plane/src/drone/executor.rs

```diff
@@ -53,13 +61,94 @@ impl<R: Runtime> Executor<R> {
+    // terminate old backends.
+    async fn terminate_preexisting_backends(
+        runtime: Arc<R>,
+        state_store: Arc<Mutex<StateStore>>,
+    ) -> Result<()> {
+        let backends = state_store
+            .lock()
+            .expect("State store lock poisoned.")
+            .active_backends()?;
+
+        if !backends.is_empty() {
+            tracing::info!(?backends, "Terminating preexisting backends");
+        }
+        let mut tasks = vec![];
+        for (backend_id, state) in backends {
+            let runtime = runtime.clone();
+            let state_store = state_store.clone();
+            let state = state.clone();
+            tasks.push(tokio::spawn(async move {
+                state_store
+                    .lock()
+                    .expect("State store lock poisoned.")
+                    .register_event(
+                        &backend_id,
+                        &state.to_terminating(TerminationKind::Hard, TerminationReason::KeyExpired),
+                        Utc::now(),
+                    )
+                    .unwrap_or_else(|_| {
+                        panic!(
+                            "Failed to register backend terminating for backend {:?}",
+                            backend_id
+                        )
+                    });
+
+                let mut backoff = ExponentialBackoff::default();
+                let mut success = false;
+                for attempt in 1..=10 {
+                    match runtime.terminate(&backend_id, true).await {
+                        Ok(()) => {
+                            success = true;
+                            break;
+                        }
+                        Err(err) => {
+                            tracing::warn!(
+                                ?err,
+                                ?backend_id,
+                                ?attempt,
+                                "Attempt failed to terminate backend"
+                            );
+                            backoff.wait().await;
+                        }
+                    }
+                }
+                if !success {
+                    tracing::warn!(
+                        ?backend_id,
+                        "Failed to terminate backend after 10 attempts. Marking terminated anyways."
+                    );
+                }
+                state_store
+                    .lock()
+                    .expect("State store lock poisoned.")
+                    .register_event(&backend_id, &state.to_terminated(None), Utc::now())
+                    .unwrap_or_else(|_| {
+                        panic!(
+                            "Failed to register backend termination for backend {:?}",
+                            backend_id
+                        )
+                    });
+            }));
+        }
+
+        for task in tasks {
+            task.await?
+        }
```

Comment:
I think this will actually run the tasks in series rather than parallel. The reason for this is that tasks are not scheduled on the executor until they are awaited, so only one task will get executor time at a time. (This is admittedly kind of unintuitive!)

Instead, `futures_util::join_all(task)` should work here. ([ref](https://docs.rs/futures-util/0.3.21/futures_util/future/fn.join_all.html)). We already have `futures_util` as a dependency.

Reply:
Actually, @rolyatmax pointed out that the `tokio::spawn` here _does_ put the task on the executor, so I was wrong. That said, I think you could avoid the `tokio::spawn` call entirely by just putting the futures in the `tasks` vec and using `join_all`. I'll leave it up to you though.